[openimageio] static linking failure

### x64-windows (dynamic)
Linking dynamically on Windows works as expected.

### x64-windows-static
When I build [openimageio] for x64-windows-static triplet my application fails to link and I get a bunch of errors like this:

`[build] OpenImageIO.lib(imageinput.cpp.obj) : error LNK2019: unresolved external symbol "void __cdecl boost::detail::set_tss_data(void const *,void (__cdecl*)(void (__cdecl*)(void *),void *),void (__cdecl*)(void *),void *,bool)" (?set_tss_data@detail@boost@@YAXPEBXP6AXP6AXPEAX@Z1@Z21_N@Z) referenced in function "public: __cdecl boost::thread_specific_ptr<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > >::~thread_specific_ptr<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > >(void)" (??1?$thread_specific_ptr@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@boost@@QEAA@XZ)`

Unless I manually add
`find_package(Boost REQUIRED COMPONENTS filesystem thread)`
to my main CMakeLists.txt

And bunch of errors like this:
`[build] OpenImageIO.lib(formatspec.cpp.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MD_DynamicRelease' in main.cpp.obj`

Unless I add
`set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} /MT")`
to my main CMakeLists.txt

### x64-linux (static)
Also require adding
`find_package(Boost REQUIRED COMPONENTS filesystem thread)`
to my main CMakeLists.txt

### x64-linux-dynamic
Linking dynamically works as expected.

Minimal repro
https://github.com/AlexeySmolenchuk/vcpkg_repro/tree/test_static_linkage

### Proposed solution
I believe it's the responsibility of individual ports to promote their dependencies and required flags.


### Additional context

For Windows I use
Microsoft Visual Studio/2019/BuildTools/VC/Tools/MSVC/14.29.30133

Linux tests in WSL:Debian
GCC 10.2.1

Another observation is
when I build [openimageio] with [opencolorio] support I get this error for **x64-linux** triplet
```
[build] /usr/bin/ld: cannot find -llzma
[build] /usr/bin/ld: cannot find -lssl
[build] /usr/bin/ld: cannot find -lcrypto
[build] collect2: error: ld returned 1 exit status
```
Maybe it's an issue with the [opencolorio] port itself. It doesn't switch flags/dependencies for derived ports.

Repro here:
https://github.com/AlexeySmolenchuk/vcpkg_repro/tree/static_linkage_with_ocio

> Another observation is when I build [openimageio] with [opencolorio] support I get this error for **x64-linux** triplet
> 
> ```
> [build] /usr/bin/ld: cannot find -llzma
> [build] /usr/bin/ld: cannot find -lssl
> [build] /usr/bin/ld: cannot find -lcrypto
> [build] collect2: error: ld returned 1 exit status
> ```
> 
> Maybe it's an issue with the [opencolorio] port itself. It doesn't switch flags/dependencies for derived ports.
> 
> Repro here: https://github.com/AlexeySmolenchuk/vcpkg_repro/tree/static_linkage_with_ocio


Sorry I didn't repro your issue on x64-linux.
```
monica@monica003:/mnt/vcpkg$ ./vcpkg install openimageio[opencolorio]
Computing installation plan...
The following packages will be built and installed:
  * opencolorio:x64-linux@2.2.1#2
    openimageio[core,opencolorio]:x64-linux@2.5.12.0#1
Additional packages (*) will be modified to complete this operation.
Detecting compiler hash for triplet x64-linux...
Compiler found: /usr/bin/c++
Restored 0 package(s) from /home/monica/.cache/vcpkg/archives in 24.6 us. Use --debug to see more details.

...

Installing 10/12 opencolorio:x64-linux@2.2.1#2...
Building opencolorio:x64-linux@2.2.1#2...
-- Downloading https://github.com/AcademySoftwareFoundation/OpenColorIO/archive/v2.2.1.tar.gz -> AcademySoftwareFoundation-OpenColorIO-v2.2.1.tar.gz...
-- Extracting source /mnt/vcpkg/downloads/AcademySoftwareFoundation-OpenColorIO-v2.2.1.tar.gz
-- Applying patch fix-dependency.patch
-- Applying patch fix-del-install-file.patch
-- Applying patch fix-func-param.patch
-- Applying patch fix-pkgconfig.patch
-- Using source at /mnt/vcpkg/buildtrees/opencolorio/src/v2.2.1-4291f52d56.clean
-- Configuring x64-linux
-- Building x64-linux-dbg
-- Building x64-linux-rel
-- Fixing pkgconfig file: /mnt/vcpkg/packages/opencolorio_x64-linux/lib/pkgconfig/OpenColorIO.pc
-- Fixing pkgconfig file: /mnt/vcpkg/packages/opencolorio_x64-linux/debug/lib/pkgconfig/OpenColorIO.pc
-- Installing: /mnt/vcpkg/packages/opencolorio_x64-linux/share/opencolorio/copyright
-- Performing post-build validation
Failed to store binary cache /home/monica/.cache/vcpkg/archives/51/51c44cecd19fcb9ae50ce8f4bc8c4ab18066a40c71fdb047e1f3401a4d929aab.zip
No such file or directory
Stored binaries in 0 destinations in 5.1 s.
Elapsed time to handle opencolorio:x64-linux: 1.7 min
opencolorio:x64-linux package ABI: 51c44cecd19fcb9ae50ce8f4bc8c4ab18066a40c71fdb047e1f3401a4d929aab

...

Installing 12/12 openimageio[core,opencolorio]:x64-linux@2.5.12.0#1...
Building openimageio[core,opencolorio]:x64-linux@2.5.12.0#1...
-- Downloading https://github.com/AcademySoftwareFoundation/OpenImageIO/archive/v2.5.12.0.tar.gz -> AcademySoftwareFoundation-OpenImageIO-v2.5.12.0.tar.gz...
-- Extracting source /mnt/vcpkg/downloads/AcademySoftwareFoundation-OpenImageIO-v2.5.12.0.tar.gz
-- Applying patch fix-dependencies.patch
-- Applying patch fix-static-ffmpeg.patch
-- Applying patch fix-openexr-dll.patch
-- Applying patch imath-version-guard.patch
-- Applying patch fix-openimageio_include_dir.patch
-- Applying patch fix-openexr-target-missing.patch
-- Applying patch fix-dependency-libraw.patch
-- Using source at /mnt/vcpkg/buildtrees/openimageio/src/v2.5.12.0-fa392fe137.clean
-- Configuring x64-linux
-- Building x64-linux-dbg
-- Building x64-linux-rel
-- Fixing pkgconfig file: /mnt/vcpkg/packages/openimageio_x64-linux/lib/pkgconfig/OpenImageIO.pc
-- Fixing pkgconfig file: /mnt/vcpkg/packages/openimageio_x64-linux/debug/lib/pkgconfig/OpenImageIO.pc
-- Installing: /mnt/vcpkg/packages/openimageio_x64-linux/share/openimageio/usage
-- Installing: /mnt/vcpkg/packages/openimageio_x64-linux/share/openimageio/copyright
-- Performing post-build validation
Failed to store binary cache /home/monica/.cache/vcpkg/archives/82/8218b4eaebcf8384ccc2e8d278cd032e6a0cf0e026429ce1576ee4d97c877374.zip
No such file or directory
Stored binaries in 0 destinations in 10 s.
Elapsed time to handle openimageio:x64-linux: 2.6 min
openimageio:x64-linux package ABI: 8218b4eaebcf8384ccc2e8d278cd032e6a0cf0e026429ce1576ee4d97c877374
Total install time: 5.4 min
The package openimageio provides CMake targets:

    find_package(OpenImageIO CONFIG REQUIRED)
    target_link_libraries(main PRIVATE OpenImageIO::OpenImageIO)
```

It is probably not a good idea to mix Windows CRT issues with Linux regular lib issues...

> ### x64-windows-static
> 
> When I build [openimageio] for x64-windows-static triplet my application fails to link [...]
> 
> And bunch of errors like this: `[build] OpenImageIO.lib(formatspec.cpp.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MD_DynamicRelease' in main.cpp.obj`
> 
> Unless I add `set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} /MT")` to my main CMakeLists.txt

`x64-windows-static` is for static CRT linkage (`OpenImageIO.lib` ... `MT_StaticRelease`). AFAICT the error message clear indicates that your app is configured for dynamic CRT linkage (`'MD_DynamicRelease' in main.cpp.obj`). Either you want to use `x64-windows-static-md` instead, or you must indeed change your project's CRT linkage configuration.

--- a/ports/openimageio/fix-dependencies.patch
+++ b/ports/openimageio/fix-dependencies.patch
@@ -1,10 +1,11 @@
 diff -u -r a/src/cmake/Config.cmake.in b/src/cmake/Config.cmake.in
 --- a/src/cmake/Config.cmake.in
 +++ b/src/cmake/Config.cmake.in
-@@ -6,6 +6,26 @@
+@@ -6,6 +6,27 @@
  
  include(CMakeFindDependencyMacro)
  
++find_dependency(Boost COMPONENTS @Boost_COMPONENTS@)
 +if(@USE_GIF@)
 +    find_dependency(GIF)
 +endif()

--- a/ports/openimageio/vcpkg.json
+++ b/ports/openimageio/vcpkg.json
@@ -1,7 +1,7 @@
 {
   "name": "openimageio",
   "version": "2.5.12.0",
-  "port-version": 1,
+  "port-version": 2,
   "description": "A library for reading and writing images, and a bunch of related classes, utilities, and application.",
   "homepage": "https://github.com/OpenImageIO/oiio",
   "license": "BSD-3-Clause",

--- a/scripts/test_ports/vcpkg-ci-openimageio/portfile.cmake
+++ b/scripts/test_ports/vcpkg-ci-openimageio/portfile.cmake
@@ -1,1 +1,4 @@
 set(VCPKG_POLICY_EMPTY_PACKAGE enabled)
+
+vcpkg_cmake_configure(SOURCE_PATH "${CURRENT_PORT_DIR}/project")
+vcpkg_cmake_build()

--- a/scripts/test_ports/vcpkg-ci-openimageio/project/CMakeLists.txt
+++ b/scripts/test_ports/vcpkg-ci-openimageio/project/CMakeLists.txt
@@ -1,0 +1,7 @@
+cmake_minimum_required(VERSION 3.7)
+project(vcpkg-ci-openimageio CXX)
+
+add_executable(main main.cpp)
+
+find_package(OpenImageIO CONFIG REQUIRED)
+target_link_libraries(main PRIVATE OpenImageIO::OpenImageIO)

--- a/scripts/test_ports/vcpkg-ci-openimageio/project/main.cpp
+++ b/scripts/test_ports/vcpkg-ci-openimageio/project/main.cpp
@@ -1,0 +1,7 @@
+#include <OpenImageIO/imageio.h>
+
+int main(int, char**)
+{
+    auto inp = OIIO::ImageInput::open("none");
+    return inp ? 0 : 1;
+}

--- a/scripts/test_ports/vcpkg-ci-openimageio/vcpkg.json
+++ b/scripts/test_ports/vcpkg-ci-openimageio/vcpkg.json
@@ -1,25 +1,44 @@
 {
   "name": "vcpkg-ci-openimageio",
-  "version": "2",
-  "description": "OpenImageIO features testing within CI.",
+  "version-string": "ci",
+  "description": "OpenImageIO testing within CI.",
   "license": "MIT",
   "dependencies": [
     {
       "name": "openimageio",
-      "features": [
-        "ffmpeg",
-        "freetype",
-        "gif",
-        "libraw",
-        "opencv",
-        "openjpeg",
+      "default-features": false
+    },
+    {
+      "name": "vcpkg-cmake",
+      "host": true
+    }
+  ],
+  "default-features": [
+    "openimageio-features"
+  ],
+  "features": {
+    "openimageio-features": {
+      "description": "Enable OpenImageIO features",
+      "dependencies": [
         {
-          "name": "pybind11",
-          "platform": "!(windows & static)"
-        },
-        "tools",
-        "webp"
+          "name": "openimageio",
+          "default-features": false,
+          "features": [
+            "ffmpeg",
+            "freetype",
+            "gif",
+            "libraw",
+            "opencv",
+            "openjpeg",
+            {
+              "name": "pybind11",
+              "platform": "!(windows & static)"
+            },
+            "tools",
+            "webp"
+          ]
+        }
       ]
     }
-  ]
+  }
 }

--- a/versions/baseline.json
+++ b/versions/baseline.json
@@ -6526,7 +6526,7 @@
     },
     "openimageio": {
       "baseline": "2.5.12.0",
-      "port-version": 1
+      "port-version": 2
     },
     "openjpeg": {
       "baseline": "2.5.2",

--- a/versions/o-/openimageio.json
+++ b/versions/o-/openimageio.json
@@ -1,6 +1,11 @@
 {
   "versions": [
     {
+      "git-tree": "e0caa7c7cee2cc24b441cc7dba90b8b183a757b0",
+      "version": "2.5.12.0",
+      "port-version": 2
+    },
+    {
       "git-tree": "ce78ee5a41eda148167cad88a553e111b40d4840",
       "version": "2.5.12.0",
       "port-version": 1

